tooling/lsp/src/requests/completion/auto_import.rs

```diff
@@ -130,66 +154,69 @@ fn module_def_id_to_reference_id(module_def_id: ModuleDefId) -> ReferenceId {
+    let mut is_relative = false;
 
-    let crate_id = module_id.krate;
-    let crate_name = match crate_id {
-        CrateId::Root(_) => Some("crate".to_string()),
-        CrateId::Crate(_) => dependencies
-            .iter()
-            .find(|dep| dep.crate_id == crate_id)
-            .map(|dep| format!("{}", dep.name)),
-        CrateId::Stdlib(_) => Some("std".to_string()),
-        CrateId::Dummy => None,
-    };
+    if let Some(module_attributes) = interner.try_module_attributes(&module_id) {
+        segments.push(&module_attributes.name);
 
-    let wrote_crate = if let Some(crate_name) = crate_name {
-        string.push_str(&crate_name);
-        true
-    } else {
-        false
-    };
+        let mut current_attributes = module_attributes;
+        loop {
+            let parent_module_id =
+                &ModuleId { krate: module_id.krate, local_id: current_attributes.parent };
 
-    let Some(module_attributes) = interner.try_module_attributes(module_id) else {
-        return string;
-    };
-
-    if wrote_crate {
-        string.push_str("::");
-    }
+            if current_module_id == parent_module_id {
+                is_relative = true;
+                break;
+            }
 
-    let mut segments = Vec::new();
-    let mut current_attributes = module_attributes;
-    loop {
-        let parent_module_id =
-            &ModuleId { krate: module_id.krate, local_id: current_attributes.parent };
-
-        // If the parent module is the current module we stop because we want a relative path to the module
-        if current_module_id == parent_module_id {
-            // When the path is relative we don't want the "crate::" prefix anymore
-            string = string.strip_prefix("crate::").unwrap_or(&string).to_string();
-            break;
-        }
+            if current_module_parent_id == Some(*parent_module_id) {
+                segments.push("super");
+                is_relative = true;
+                break;
+            }
 
-        let Some(parent_attributes) = interner.try_module_attributes(parent_module_id) else {
-            break;
-        };
+            let Some(parent_attributes) = interner.try_module_attributes(parent_module_id) else {
+                break;
+            };
 
-        segments.push(&parent_attributes.name);
-        current_attributes = parent_attributes;
+            segments.push(&parent_attributes.name);
+            current_attributes = parent_attributes;
+        }
     }
 
-    for segment in segments.iter().rev() {
-        string.push_str(segment);
-        string.push_str("::");
-    }
+    let crate_id = module_id.krate;
+    let crate_name = if is_relative {
+        None
+    } else {
+        match crate_id {
+            CrateId::Root(_) => {
+                if Some(module_id) == current_module_parent_id {
+                    Some("super".to_string())
+                } else {
+                    Some("crate".to_string())
+                }
```

Comment:
Is this ever going to trigger? If `Some(module_id) == current_module_parent_id` then we're going to have `is_relative == true` surely?

Reply:
Oh, it confused me too, but it can happen. For example if we have:

```rust
// We are in module foo
mod tests {
  // We want to import something that's in module foo
}
```

then from `tests` we get the parent, which is `foo`, and we want to get the full path of something in `foo` from `foo`. That can be `super`. So if there's `fn bar()` inside `foo` we end up using `super::bar`.

Maybe there's a way to make this function clearer?

Reply:
Yeah, this is nice now.